File: pms-backend/src/main/java/com/pms/service/EmployeeDetailsService.java
package com.pms.service;

import com.pms.dto.EmployeeDetailsRequest;
import com.pms.dto.EmployeeDetailsResponse;
import com.pms.entity.EmployeeDetails;
import com.pms.repository.EmployeeDetailsRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class EmployeeDetailsService {

    private final EmployeeDetailsRepository employeeDetailsRepository;

    public List<EmployeeDetailsResponse> getAllEmployeeDetails() {
        log.info("Fetching all employee details");
        return employeeDetailsRepository.findAll()
                .stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    public EmployeeDetailsResponse getEmployeeDetailsByCode(String code) {
        log.info("Fetching employee details for code: {}", code);
        return employeeDetailsRepository.findById(code)
                .map(this::mapToResponse)
                .orElseThrow(() -> new RuntimeException("Employee not found with code: " + code));
    }

    public EmployeeDetailsResponse createEmployeeDetails(EmployeeDetailsRequest request) {
        log.info("Creating employee details for code: {}", request.getEmployeeCode());
        
        EmployeeDetails employeeDetails = EmployeeDetails.builder()
                .employeeCode(request.getEmployeeCode())
                .name(request.getName())
                .presentDesignationFullName(request.getPresentDesignationFullName())
                .centre(request.getCentre())
                .userId(request.getUserId())
                .regStatus(request.getRegStatus())
                .regTime(LocalDate.now())
                .build();

        EmployeeDetails saved = employeeDetailsRepository.save(employeeDetails);
        log.info("Employee details created successfully: {}", saved.getEmployeeCode());
        return mapToResponse(saved);
    }

    public EmployeeDetailsResponse updateEmployeeDetails(String code, EmployeeDetailsRequest request) {
        log.info("Updating employee details for code: {}", code);
        
        EmployeeDetails existing = employeeDetailsRepository.findById(code)
                .orElseThrow(() -> new RuntimeException("Employee not found with code: " + code));

        existing.setName(request.getName());
        existing.setPresentDesignationFullName(request.getPresentDesignationFullName());
        existing.setCentre(request.getCentre());
        existing.setUserId(request.getUserId());
        existing.setRegStatus(request.getRegStatus());

        EmployeeDetails updated = employeeDetailsRepository.save(existing);
        log.info("Employee details updated successfully: {}", updated.getEmployeeCode());
        return mapToResponse(updated);
    }

    public void deleteEmployeeDetails(String code) {
        log.info("Deleting employee details for code: {}", code);
        if (!employeeDetailsRepository.existsById(code)) {
            throw new RuntimeException("Employee not found with code: " + code);
        }
        employeeDetailsRepository.deleteById(code);
        log.info("Employee details deleted successfully: {}", code);
    }

    public List<EmployeeDetailsResponse> getEmployeeDetailsByStatus(String status) {
        log.info("Fetching employee details by status: {}", status);
        return employeeDetailsRepository.findByRegStatus(status)
                .stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    public List<EmployeeDetailsResponse> getEmployeeDetailsByCentre(String centre) {
        log.info("Fetching employee details by centre: {}", centre);
        return employeeDetailsRepository.findByCentre(centre)
                .stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    public void deactivateEmployeeDetails(String code) {
        log.info("Deactivating employee details for code: {}", code);
        EmployeeDetails existing = employeeDetailsRepository.findById(code)
                .orElseThrow(() -> new RuntimeException("Employee not found with code: " + code));
        existing.setRegStatus("I");
        employeeDetailsRepository.save(existing);
        log.info("Employee details deactivated successfully: {}", code);
    }

    private EmployeeDetailsResponse mapToResponse(EmployeeDetails entity) {
        return EmployeeDetailsResponse.builder()
                .employeeCode(entity.getEmployeeCode())
                .name(entity.getName())
                .presentDesignationFullName(entity.getPresentDesignationFullName())
                .centre(entity.getCentre())
                .userId(entity.getUserId())
                .regStatus(entity.getRegStatus())
                .regTime(entity.getRegTime() != null ? entity.getRegTime().toString() : null)
                .build();
    }
}

File: src/services/employeeDetailsService.ts
const API_BASE_URL = "http://localhost:7080/api";

export interface EmployeeDetailsRequest {
  employeeCode: string;
  name: string;
  presentDesignationFullName: string;
  centre: string;
  userId: string;
  regStatus: string;
  regTime?: string;
}

export interface EmployeeDetailsResponse {
  employeeCode: string;
  name: string;
  presentDesignationFullName: string;
  centre: string;
  userId: string;
  regStatus: string;
  regTime?: string;
}

class EmployeeDetailsService {
  async getAllEmployeeDetails(): Promise<EmployeeDetailsResponse[]> {
    const response = await fetch(`${API_BASE_URL}/employee-details`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      const error = await response.text();
      console.error("Failed to fetch employee details:", error);
      throw new Error("Failed to fetch employee details");
    }

    return await response.json();
  }

  async getEmployeeDetailsByCode(code: string): Promise<EmployeeDetailsResponse> {
    const response = await fetch(`${API_BASE_URL}/employee-details/${code}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      throw new Error("Failed to fetch employee details");
    }

    return await response.json();
  }

  async createEmployeeDetails(request: EmployeeDetailsRequest): Promise<EmployeeDetailsResponse> {
    const response = await fetch(`${API_BASE_URL}/employee-details`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      const error = await response.text();
      console.error("Failed to create employee details:", error);
      throw new Error("Failed to create employee details");
    }

    return await response.json();
  }

  async updateEmployeeDetails(
    code: string,
    request: EmployeeDetailsRequest
  ): Promise<EmployeeDetailsResponse> {
    const response = await fetch(`${API_BASE_URL}/employee-details/${code}`, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      const error = await response.text();
      console.error("Failed to update employee details:", error);
      throw new Error("Failed to update employee details");
    }

    return await response.json();
  }

  async deleteEmployeeDetails(code: string): Promise<void> {
    const response = await fetch(`${API_BASE_URL}/employee-details/${code}`, {
      method: "DELETE",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      const error = await response.text();
      console.error("Failed to delete employee details:", error);
      throw new Error("Failed to delete employee details");
    }
  }

  async getEmployeeDetailsByStatus(status: string): Promise<EmployeeDetailsResponse[]> {
    const response = await fetch(`${API_BASE_URL}/employee-details/by-status/${status}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      throw new Error("Failed to fetch employee details by status");
    }

    return await response.json();
  }

  async getEmployeeDetailsByCentre(centre: string): Promise<EmployeeDetailsResponse[]> {
    const response = await fetch(`${API_BASE_URL}/employee-details/by-centre/${centre}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      throw new Error("Failed to fetch employee details by centre");
    }

    return await response.json();
  }

  async deactivateEmployeeDetails(code: string): Promise<void> {
    const response = await fetch(`${API_BASE_URL}/employee-details/${code}/deactivate`, {
      method: "PATCH",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      const error = await response.text();
      console.error("Failed to deactivate employee details:", error);
      throw new Error("Failed to deactivate employee details");
    }
  }
}

export const employeeDetailsService = new EmployeeDetailsService();
